src/distinct.rs

```diff
@@ -227,13 +182,104 @@ pub fn rearrange_sequence(colors: &mut Vec<Color>, metric: DistanceMetric) {
         DistanceMetric::CIEDE2000 => c1.distance_delta_e_ciede2000(c2),
     };
 
+    // vector where the i-th element contains the minimum distance to the colors from 0 to i-1.
+    let mut min_distances = vec![i32::max_value(); colors.len()];
+
     for i in 1..colors.len() {
-        let (left, right) = colors.split_at_mut(i);
-        right.sort_unstable_by_key(|c| {
-            left.iter()
-                .map(|c2| (-distance(c2, c) * 1000.0) as i32)
-                .max()
-        });
+        let mut max_i = colors.len();
+        let mut max_d = i32::min_value();
+
+        for j in i..colors.len() {
+            min_distances[j] =
+                min_distances[j].min((distance(&colors[j], &colors[i - 1]) * 1000.0) as i32);
+
+            if min_distances[j] > max_d {
+                max_i = j;
+                max_d = min_distances[j];
+            }
+        }
+
+        colors.swap(i, max_i);
+        min_distances.swap(i, max_i);
+    }
+}
+
+impl DistanceResult {
+    fn new(colors: &[(Color, Lab)], distance_metric: DistanceMetric) -> Self {
+        let mut result = DistanceResult {
+            closest_distances: vec![(scalar::MAX, std::usize::MAX); colors.len()],
+            closest_pair: (std::usize::MAX, std::usize::MAX),
+            mean_closest_distance: 0.0,
+            min_closest_distance: scalar::MAX,
+            distance_metric,
+        };
+
+        for i in 0..colors.len() {
+            result.update_distances(colors, i, false);
+        }
+        result.update_totals();
+
+        result
+    }
+
+    fn update(&self, colors: &[(Color, Lab)], changed_color: usize) -> Self {
+        let mut result = self.clone();
+        result.update_distances(colors, changed_color, true);
+        result.update_totals();
+        result
+    }
+
+    fn update_distances(&mut self, colors: &[(Color, Lab)], color: usize, changed: bool) {
+        self.closest_distances[color] = (scalar::MAX, std::usize::MAX);
+
+        // we need to recalculate distances for nodes where the previous min dist was with
+        // changed_color but it's not anymore (potentially).
+        let mut to_recalc = Vec::with_capacity(colors.len());
+
+        for (i, c) in colors.iter().enumerate() {
+            if i == color {
+                continue;
+            }
+
+            let dist = self.distance(c, &colors[color]);
+
+            if dist < self.closest_distances[i].0 {
+                self.closest_distances[i] = (dist, color);
+                self.closest_distances[color] = (dist, i);
```

Comment:
I think there is a mistake here: If `dist < self.closest_distances[i].0`, it is okay to update `self.closest_distances[i]`. But that doesn't necessarily mean that `self.closest_distances[color]` should also be updated, right? `color` could be closer to another (third) color but still be the nearest neighbor for `i`.

I think there should be a separate `if` block for updating `self.closest_distances[color]`. This could also be the reason for `scalar::MAX` values appearing in the mean computation. Imagine that we shift `color` very far away from all other colors and then recompute the distances. Since it is not a nearest neighbor of any of the other colors, `self.closest_distances[color]` will never be updated in this loop and the value will still be `scalar::MAX`.

Reply:
you're absolutely right, my bad I didn't catch the error :sweat_smile: 

I've added the separate check and it seems that the D_min is now as high as before. Occasionally I get some lower results like 98.3 but it seems to me that the current version of pastel also returns those values sometimes.

Reply:
> Occasionally I get some lower results like 98.3 but it seems to me that the current version of pastel also returns those values sometimes.

I'm sure that's possible. The really nice thing about your PR is that we can now probably adapt the simulated annealing parameters a bit because it is now so fast. I would like to increase the iteration counts (and simultaneously decrease the cooling rates). This should lead to much more stable results.

Reply:
yeah, I didn't want to change the parameters because I think the plan is too actually expose those parameters as command line options, but I guess it's ok to tune the parameters now.

However, I think it would be interesting to have an "adaptive" metric in the simulated annealing process too that would potentially stop the process early if, say, the relative error is negligible or something.

Reply:
> However, I think it would be interesting to have an "adaptive" metric in the simulated annealing process too that would potentially stop the process early if, say, the relative error is negligible or something.

Absolutely